File: bindings/julia/Seatuya.jl
# Seatuya.jl — Julia FFI bindings for libseatuya
#
# Pure Julia module using the built-in `ccall` for every function
# in libseatuya.  No external dependencies.
#
# Usage:
#   using Seatuya
#   dev = Seatuya.create(device_id, "192.168.1.100", local_key, "3.4")
#   println(Seatuya.turn_on(dev, 1))
#   println(Seatuya.status(dev))
#   Seatuya.destroy(dev)

module Seatuya

export version, create, alloc, destroy,
       set_credentials, get_device_id, get_local_key, get_ip,
       connect, disconnect, is_connected, reconnect,
       set_retry_limit, set_retry_delay, get_retry_limit, get_retry_delay,
       negotiate_session, negotiate_session_start, negotiate_session_finalize,
       get_protocol, get_session_state, get_socket_state, get_last_error,
       set_async_mode, is_socket_readable, is_socket_writable, set_session_ready,
       build_message, decode_message, generate_payload, send_frame, receive_frame,
       set_value, turn_on, turn_off, status, heartbeat,
       set_device22, is_device22,
       Command, Protocol, SessionState, SocketState,
       DEFAULT_PORT, BUFSIZE, DEFAULT_RETRY_LIMIT, DEFAULT_RETRY_DELAY_MS

# --- Library path ---
const lib = if haskey(ENV, "SEATUYA_LIB")
    ENV["SEATUYA_LIB"]
elseif Sys.isapple()
    "libseatuya.dylib"
elseif Sys.iswindows()
    "seatuya.dll"
else
    "libseatuya.so"
end

# --- Enums ---
@enum TuyaCommand::Cint begin
    CMD_UDP = 0
    CMD_AP_CONFIG = 1
    CMD_ACTIVE = 2
    CMD_BIND = 3
    CMD_RENAME_GW = 4
    CMD_RENAME_DEVICE = 5
    CMD_UNBIND = 6
    CMD_CONTROL = 7
    CMD_STATUS = 8
    CMD_HEART_BEAT = 9
    CMD_DP_QUERY = 10
    CMD_QUERY_WIFI = 11
    CMD_TOKEN_BIND = 12
    CMD_CONTROL_NEW = 13
    CMD_ENABLE_WIFI = 14
    CMD_DP_QUERY_NEW = 16
    CMD_SCENE_EXECUTE = 17
    CMD_UPDATEDPS = 18
    CMD_UDP_NEW = 19
    CMD_AP_CONFIG_NEW = 20
    CMD_GET_LOCAL_TIME = 28
    CMD_WEATHER_OPEN = 32
    CMD_WEATHER_DATA = 33
    CMD_STATE_UPLOAD_SYN = 34
    CMD_STATE_UPLOAD_SYN_RECV = 35
    CMD_HEART_BEAT_STOP = 37
    CMD_STREAM_TRANS = 38
    CMD_GET_WIFI_STATUS = 43
    CMD_WIFI_CONNECT_TEST = 44
    CMD_GET_MAC = 45
    CMD_GET_IR_STATUS = 46
    CMD_IR_TX_RX_TEST = 47
    CMD_LAN_GW_ACTIVE = 240
    CMD_LAN_SUB_DEV_REQUEST = 241
    CMD_LAN_DELETE_SUB_DEV = 242
    CMD_LAN_REPORT_SUB_DEV = 243
    CMD_LAN_SCENE = 244
    CMD_LAN_PUBLISH_CLOUD_CONFIG = 245
    CMD_LAN_PUBLISH_APP_CONFIG = 246
    CMD_LAN_EXPORT_APP_CONFIG = 247
    CMD_LAN_PUBLISH_SCENE_PANEL = 248
    CMD_LAN_REMOVE_GW = 249
    CMD_LAN_CHECK_GW_UPDATE = 250
    CMD_LAN_GW_UPDATE = 251
    CMD_LAN_SET_GW_CHANNEL = 252
end

@enum TuyaProtocol::Cint PROTO_V31=0 PROTO_V33=1 PROTO_V34=2 PROTO_V35=3
@enum SessionState::Cint SESSION_INVALID=0 SESSION_STARTING=1 SESSION_FINALIZING=2 SESSION_ESTABLISHED=3
@enum SocketState::Cint SOCK_NO_SUCH_HOST=0 SOCK_NO_SOCK_AVAIL=1 SOCK_FAILED=2 SOCK_DISCONNECTED=3 SOCK_CONNECTING=4 SOCK_CONNECTED=5 SOCK_READY=6 SOCK_RECEIVING=7

# Export-friendly aliases
const Command = TuyaCommand; const Protocol = TuyaProtocol
const DEFAULT_PORT = 6668; const BUFSIZE = 1024
const DEFAULT_RETRY_LIMIT = 5; const DEFAULT_RETRY_DELAY_MS = 100

# --- Lifecycle ---
version() = unsafe_string(ccall((:tuya_version, lib), Cstring, ()))

function create(device_id::String, address::String, local_key::String, ver::String)
    ptr = ccall((:tuya_create, lib), Ptr{Cvoid},
                (Cstring, Cstring, Cstring, Cstring),
                device_id, address, local_key, ver)
    return ptr == C_NULL ? nothing : ptr
end

function alloc(ver::String)
    ptr = ccall((:tuya_alloc, lib), Ptr{Cvoid}, (Cstring,), ver)
    return ptr == C_NULL ? nothing : ptr
end

destroy(dev) = ccall((:tuya_destroy, lib), Cvoid, (Ptr{Cvoid},), dev)

# --- Credentials ---
function set_credentials(dev, device_id::String, local_key::String)
    ccall((:tuya_set_credentials, lib), Cvoid, (Ptr{Cvoid}, Cstring, Cstring), dev, device_id, local_key)
end
get_device_id(dev) = unsafe_string(ccall((:tuya_get_device_id, lib), Cstring, (Ptr{Cvoid},), dev))
get_local_key(dev) = unsafe_string(ccall((:tuya_get_local_key, lib), Cstring, (Ptr{Cvoid},), dev))
get_ip(dev) = unsafe_string(ccall((:tuya_get_ip, lib), Cstring, (Ptr{Cvoid},), dev))

# --- Connection ---
connect(dev, hostname::String) = ccall((:tuya_connect, lib), Bool, (Ptr{Cvoid}, Cstring), dev, hostname)
disconnect(dev) = ccall((:tuya_disconnect, lib), Cvoid, (Ptr{Cvoid},), dev)
is_connected(dev) = ccall((:tuya_is_connected, lib), Bool, (Ptr{Cvoid},), dev)
reconnect(dev) = ccall((:tuya_reconnect, lib), Bool, (Ptr{Cvoid},), dev)

# --- Retry settings ---
set_retry_limit(dev, limit::Integer) = ccall((:tuya_set_retry_limit, lib), Cvoid, (Ptr{Cvoid}, Cint), dev, limit)
set_retry_delay(dev, ms::Integer) = ccall((:tuya_set_retry_delay, lib), Cvoid, (Ptr{Cvoid}, Cint), dev, ms)
get_retry_limit(dev) = Int(ccall((:tuya_get_retry_limit, lib), Cint, (Ptr{Cvoid},), dev))
get_retry_delay(dev) = Int(ccall((:tuya_get_retry_delay, lib), Cint, (Ptr{Cvoid},), dev))

# --- Session ---
negotiate_session(dev, key::String) = ccall((:tuya_negotiate_session, lib), Bool, (Ptr{Cvoid}, Cstring), dev, key)
negotiate_session_start(dev, key::String) = ccall((:tuya_negotiate_session_start, lib), Bool, (Ptr{Cvoid}, Cstring), dev, key)
function negotiate_session_finalize(dev, buf::Vector{UInt8}, key::String)
    ccall((:tuya_negotiate_session_finalize, lib), Bool, (Ptr{Cvoid}, Ptr{Cvoid}, Cint, Cstring), dev, buf, length(buf), key)
end

# --- State queries ---
get_protocol(dev) = TuyaProtocol(ccall((:tuya_get_protocol, lib), Cint, (Ptr{Cvoid},), dev))
get_session_state(dev) = SessionState(ccall((:tuya_get_session_state, lib), Cint, (Ptr{Cvoid},), dev))
get_socket_state(dev) = SocketState(ccall((:tuya_get_socket_state, lib), Cint, (Ptr{Cvoid},), dev))
get_last_error(dev) = Int(ccall((:tuya_get_last_error, lib), Cint, (Ptr{Cvoid},), dev))

# --- Async ---
set_async_mode(dev, flag::Bool) = ccall((:tuya_set_async_mode, lib), Cvoid, (Ptr{Cvoid}, Bool), dev, flag)
is_socket_readable(dev) = ccall((:tuya_is_socket_readable, lib), Bool, (Ptr{Cvoid},), dev)
is_socket_writable(dev) = ccall((:tuya_is_socket_writable, lib), Bool, (Ptr{Cvoid},), dev)
set_session_ready(dev) = ccall((:tuya_set_session_ready, lib), Bool, (Ptr{Cvoid},), dev)

# --- Low-level ---
function build_message(dev, cmd::TuyaCommand, payload::String, key::String)
    buf = zeros(UInt8, BUFSIZE)
    n = ccall((:tuya_build_message, lib), Cint,
              (Ptr{Cvoid}, Ptr{Cvoid}, Cint, Cstring, Cstring),
              dev, buf, Int(cmd), payload, key)
    return n > 0 ? buf[1:n] : nothing
end

function decode_message(dev, buf::Vector{UInt8}, key::String)
    ptr = ccall((:tuya_decode_message, lib), Cstring,
                (Ptr{Cvoid}, Ptr{Cvoid}, Cint, Cstring),
                dev, buf, length(buf), key)
    if ptr == C_NULL; return nothing; end
    s = unsafe_string(ptr)
    ccall((:tuya_free_string, lib), Cvoid, (Cstring,), ptr)
    return s
end

function generate_payload(dev, cmd::TuyaCommand, device_id::String, datapoints::String="")
    ptr = ccall((:tuya_generate_payload, lib), Cstring,
                (Ptr{Cvoid}, Cint, Cstring, Cstring),
                dev, Int(cmd), device_id, datapoints)
    if ptr == C_NULL; return nothing; end
    s = unsafe_string(ptr)
    ccall((:tuya_free_string, lib), Cvoid, (Cstring,), ptr)
    return s
end

send_frame(dev, buf::Vector{UInt8}) = Int(ccall((:tuya_send, lib), Cint, (Ptr{Cvoid}, Ptr{Cvoid}, Cint), dev, buf, length(buf)))

function receive_frame(dev, maxsize::Integer=BUFSIZE, minsize::Integer=0)
    buf = zeros(UInt8, maxsize)
    n = ccall((:tuya_receive, lib), Cint, (Ptr{Cvoid}, Ptr{Cvoid}, Cint, Cint), dev, buf, maxsize, minsize)
    return n > 0 ? buf[1:n] : nothing
end

# --- High-level round-trip ---
function _consume(ptr::Ptr{Cvoid})
    if ptr == C_NULL; return nothing; end
    s = unsafe_string(ptr)
    ccall((:tuya_free_string, lib), Cvoid, (Cstring,), ptr)
    return s
end

function set_value(dev, dp::Integer, value)
    ptr = if value isa Bool
        ccall((:tuya_set_value_bool, lib), Cstring, (Ptr{Cvoid}, Cint, Bool), dev, dp, value)
    elseif value isa Integer
        ccall((:tuya_set_value_int, lib), Cstring, (Ptr{Cvoid}, Cint, Cint), dev, dp, value)
    elseif value isa AbstractFloat
        ccall((:tuya_set_value_float, lib), Cstring, (Ptr{Cvoid}, Cint, Float64), dev, dp, value)
    else
        ccall((:tuya_set_value_string, lib), Cstring, (Ptr{Cvoid}, Cint, Cstring), dev, dp, string(value))
    end
    return _consume(ptr)
end

turn_on(dev, switch_dp::Integer=1) = _consume(ccall((:tuya_turn_on, lib), Cstring, (Ptr{Cvoid}, Cint), dev, switch_dp))
turn_off(dev, switch_dp::Integer=1) = _consume(ccall((:tuya_turn_off, lib), Cstring, (Ptr{Cvoid}, Cint), dev, switch_dp))
status(dev) = _consume(ccall((:tuya_status, lib), Cstring, (Ptr{Cvoid},), dev))
heartbeat(dev) = _consume(ccall((:tuya_heartbeat, lib), Cstring, (Ptr{Cvoid},), dev))

# --- device22 ---
set_device22(dev, null_dps_json::String) = ccall((:tuya_set_device22, lib), Cvoid, (Ptr{Cvoid}, Cstring), dev, null_dps_json)
is_device22(dev) = ccall((:tuya_is_device22, lib), Bool, (Ptr{Cvoid},), dev)

end # module Seatuya

File: bindings/haskell/Seatuya.hs
-- Seatuya.hs -- Haskell FFI bindings for libseatuya
--
-- Uses Foreign.Ptr, Foreign.C, and Foreign.Storable for C interop.
-- The opaque tuya_device_t* is wrapped in a ForeignPtr for GC safety.
-- Set SEATUYA_LIB to a custom library path, or use LD_LIBRARY_PATH.
--
-- Usage:
--   import Seatuya
--   dev <- create "id" "192.168.1.100" "key" "3.4"
--   turnOn dev 1 >>= putStrLn
--   destroy dev

{-# LANGUAGE ForeignFunctionInterface #-}
module Seatuya
    ( TuyaDevice
    , version, create, alloc, destroy
    , setCredentials, getDeviceId, getLocalKey, getIp
    , connect, disconnect, isConnected, reconnect
    , setRetryLimit, setRetryDelay, getRetryLimit, getRetryDelay
    , negotiateSession, negotiateSessionStart, negotiateSessionFinalize
    , getProtocol, getSessionState, getSocketState, getLastError
    , setAsyncMode, isSocketReadable, isSocketWritable, setSessionReady
    , buildMessage, decodeMessage, generatePayload, send, receive
    , setValueBool, setValueInt, setValueString, setValueFloat
    , turnOn, turnOff, status, heartbeat
    , setDevice22, isDevice22
    , setValue
    , Command(..), Protocol(..), SessionState(..), SocketState(..)
    , defaultPort, bufSize, defaultRetryLimit, defaultRetryDelay
    ) where

import Foreign
import Foreign.C.String
import Foreign.C.Types
import System.Environment (lookupEnv)
import Control.Monad (when)
import Data.Int (Int32)
import System.IO.Unsafe (unsafePerformIO)

-- ---------------------------------------------------------------------------
-- Device handle with automatic finalization
-- ---------------------------------------------------------------------------

data TuyaDevice = TuyaDevice (ForeignPtr ())

-- ---------------------------------------------------------------------------
-- Foreign imports (link-time)
-- ---------------------------------------------------------------------------

foreign import ccall "tuya_version"  c_version :: IO CString
foreign import ccall "tuya_create"   c_create  :: CString -> CString -> CString -> CString -> IO (Ptr ())
foreign import ccall "tuya_alloc"    c_alloc   :: CString -> IO (Ptr ())
foreign import ccall "tuya_destroy"  c_destroy :: Ptr () -> IO ()
foreign import ccall "&tuya_destroy" c_destroy_finalizer :: FunPtr (Ptr () -> IO ())

foreign import ccall "tuya_set_credentials" c_set_creds   :: Ptr () -> CString -> CString -> IO ()
foreign import ccall "tuya_get_device_id"   c_get_devid   :: Ptr () -> IO CString
foreign import ccall "tuya_get_local_key"   c_get_key     :: Ptr () -> IO CString
foreign import ccall "tuya_get_ip"          c_get_ip      :: Ptr () -> IO CString

foreign import ccall "tuya_connect"       c_connect       :: Ptr () -> CString -> IO CInt
foreign import ccall "tuya_disconnect"    c_disconnect    :: Ptr () -> IO ()
foreign import ccall "tuya_is_connected"  c_is_connected  :: Ptr () -> IO CInt
foreign import ccall "tuya_reconnect"     c_reconnect     :: Ptr () -> IO CInt

foreign import ccall "tuya_set_retry_limit" c_set_retry_limit :: Ptr () -> CInt -> IO ()
foreign import ccall "tuya_set_retry_delay" c_set_retry_delay :: Ptr () -> CInt -> IO ()
foreign import ccall "tuya_get_retry_limit" c_get_retry_limit :: Ptr () -> IO CInt
foreign import ccall "tuya_get_retry_delay" c_get_retry_delay :: Ptr () -> IO CInt

foreign import ccall "tuya_negotiate_session"         c_neg_sess   :: Ptr () -> CString -> IO CInt
foreign import ccall "tuya_negotiate_session_start"    c_neg_start  :: Ptr () -> CString -> IO CInt
foreign import ccall "tuya_negotiate_session_finalize" c_neg_final  :: Ptr () -> Ptr CChar -> CInt -> CString -> IO CInt

foreign import ccall "tuya_get_protocol"       c_get_proto    :: Ptr () -> IO CInt
foreign import ccall "tuya_get_session_state"  c_get_sess     :: Ptr () -> IO CInt
foreign import ccall "tuya_get_socket_state"   c_get_sock     :: Ptr () -> IO CInt
foreign import ccall "tuya_get_last_error"     c_get_err      :: Ptr () -> IO CInt

foreign import ccall "tuya_set_async_mode"       c_set_async  :: Ptr () -> CInt -> IO ()
foreign import ccall "tuya_is_socket_readable"   c_readable   :: Ptr () -> IO CInt
foreign import ccall "tuya_is_socket_writable"   c_writable   :: Ptr () -> IO CInt
foreign import ccall "tuya_set_session_ready"    c_set_ready  :: Ptr () -> IO CInt

foreign import ccall "tuya_build_message"    c_build :: Ptr () -> Ptr CChar -> CInt -> CString -> CString -> CInt
foreign import ccall "tuya_decode_message"   c_decode :: Ptr () -> Ptr CChar -> CInt -> CString -> IO CString
foreign import ccall "tuya_generate_payload" c_gen :: Ptr () -> CInt -> CString -> CString -> IO CString
foreign import ccall "tuya_send"    c_send    :: Ptr () -> Ptr CChar -> CInt -> IO CInt
foreign import ccall "tuya_receive" c_receive :: Ptr () -> Ptr CChar -> CInt -> CInt -> IO CInt

foreign import ccall "tuya_set_value_bool"   c_set_bool   :: Ptr () -> CInt -> CInt -> IO CString
foreign import ccall "tuya_set_value_int"    c_set_int    :: Ptr () -> CInt -> CInt -> IO CString
foreign import ccall "tuya_set_value_string" c_set_str    :: Ptr () -> CInt -> CString -> IO CString
foreign import ccall "tuya_set_value_float"  c_set_float  :: Ptr () -> CInt -> CDouble -> IO CString

foreign import ccall "tuya_turn_on"    c_turn_on    :: Ptr () -> CInt -> IO CString
foreign import ccall "tuya_turn_off"   c_turn_off   :: Ptr () -> CInt -> IO CString
foreign import ccall "tuya_status"     c_status     :: Ptr () -> IO CString
foreign import ccall "tuya_heartbeat"  c_heartbeat  :: Ptr () -> IO CString

foreign import ccall "tuya_free_string" c_free_str :: CString -> IO ()

foreign import ccall "tuya_set_device22" c_set_d22 :: Ptr () -> CString -> IO ()
foreign import ccall "tuya_is_device22"  c_is_d22  :: Ptr () -> IO CInt

-- ---------------------------------------------------------------------------
-- Constants
-- ---------------------------------------------------------------------------

-- | Tuya command types (43 values matching enum tuya_command).
data Command = CmdUDP | CmdApConfig | CmdActive | CmdBind | CmdRenameGw
             | CmdRenameDevice | CmdUnbind | CmdControl | CmdStatus
             | CmdHeartBeat | CmdDpQuery | CmdQueryWifi | CmdTokenBind
             | CmdControlNew | CmdEnableWifi | CmdDpQueryNew | CmdSceneExecute
             | CmdUpdateDps | CmdUdpNew | CmdApConfigNew | CmdGetLocalTime
             | CmdWeatherOpen | CmdWeatherData | CmdStateUploadSyn
             | CmdStateUploadSynRecv | CmdHeartBeatStop | CmdStreamTrans
             | CmdGetWifiStatus | CmdWifiConnectTest | CmdGetMac
             | CmdGetIrStatus | CmdIrTxRxTest | CmdLanGwActive
             | CmdLanSubDevRequest | CmdLanDeleteSubDev | CmdLanReportSubDev
             | CmdLanScene | CmdLanPublishCloudConfig | CmdLanPublishAppConfig
             | CmdLanExportAppConfig | CmdLanPublishScenePanel | CmdLanRemoveGw
             | CmdLanCheckGwUpdate | CmdLanGwUpdate | CmdLanSetGwChannel
             deriving (Show, Eq, Enum)

cmdVal :: Command -> CInt
cmdVal c = fromIntegral (fromEnum c)

cmdCtrl, cmdDpQuery, cmdHeartBeat, cmdStatus, cmdCtrlNew, cmdDpQueryNew :: CInt
cmdCtrl     = cmdVal CmdControl
cmdDpQuery  = cmdVal CmdDpQuery
cmdHeartBeat = cmdVal CmdHeartBeat
cmdStatus   = cmdVal CmdStatus
cmdCtrlNew  = cmdVal CmdControlNew
cmdDpQueryNew = cmdVal CmdDpQueryNew

-- | Protocol version.
data Protocol = ProtoV31 | ProtoV33 | ProtoV34 | ProtoV35
              deriving (Show, Eq, Enum)

protoFromInt :: CInt -> Protocol
protoFromInt 0 = ProtoV31; protoFromInt 1 = ProtoV33
protoFromInt 2 = ProtoV34; protoFromInt _ = ProtoV35

-- | Session state.
data SessionState = SessInvalid | SessStarting | SessFinalizing | SessEstablished
                  deriving (Show, Eq, Enum)

sessFromInt :: CInt -> SessionState
sessFromInt 0 = SessInvalid; sessFromInt 1 = SessStarting
sessFromInt 2 = SessFinalizing; sessFromInt _ = SessEstablished

-- | Socket state.
data SocketState = SockNoSuchHost | SockNoSockAvail | SockFailed
                 | SockDisconnected | SockConnecting | SockConnected
                 | SockReady | SockReceiving
                 deriving (Show, Eq, Enum)

sockFromInt :: CInt -> SocketState
sockFromInt 0 = SockNoSuchHost; sockFromInt 1 = SockNoSockAvail
sockFromInt 2 = SockFailed; sockFromInt 3 = SockDisconnected
sockFromInt 4 = SockConnecting; sockFromInt 5 = SockConnected
sockFromInt 6 = SockReady; sockFromInt _ = SockReceiving

defaultPort, bufSize, defaultRetryLimit, defaultRetryDelay :: Int
defaultPort = 6668
bufSize = 1024
defaultRetryLimit = 5
defaultRetryDelay = 100

-- ---------------------------------------------------------------------------
-- Helpers
-- ---------------------------------------------------------------------------

toBool :: CInt -> Bool
toBool 0 = False; toBool _ = True

consume :: CString -> IO String
consume p = do
    s <- peekCString p
    c_free_str p
    return s

withDevice :: TuyaDevice -> (Ptr () -> IO a) -> IO a
withDevice (TuyaDevice fp) f = withForeignPtr fp f

-- ---------------------------------------------------------------------------
-- Public API -- Version
-- ---------------------------------------------------------------------------

version :: IO String
version = c_version >>= peekCString

-- ---------------------------------------------------------------------------
-- Lifecycle
-- ---------------------------------------------------------------------------

create :: String -> String -> String -> String -> IO (Maybe TuyaDevice)
create did addr key ver =
    withCString did  $ \cdid ->
    withCString addr $ \caddr ->
    withCString key  $ \ckey ->
    withCString ver  $ \cver  -> do
        p <- c_create cdid caddr ckey cver
        if p == nullPtr then return Nothing
        else Just <$> (TuyaDevice <$> newForeignPtr c_destroy_finalizer p)

alloc :: String -> IO (Maybe TuyaDevice)
alloc ver = withCString ver $ \cv -> do
    p <- c_alloc cv
    if p == nullPtr then return Nothing
    else Just <$> (TuyaDevice <$> newForeignPtr c_destroy_finalizer p)

destroy :: TuyaDevice -> IO ()
destroy (TuyaDevice fp) = withForeignPtr fp $ \p -> c_destroy p

-- ---------------------------------------------------------------------------
-- Credentials
-- ---------------------------------------------------------------------------

setCredentials :: TuyaDevice -> String -> String -> IO ()
setCredentials dev did key = withDevice dev $ \d ->
    withCString did $ \cd -> withCString key $ \ck -> c_set_creds d cd ck

getDeviceId :: TuyaDevice -> IO String
getDeviceId dev = withDevice dev $ \d -> c_get_devid d >>= peekCString

getLocalKey :: TuyaDevice -> IO String
getLocalKey dev = withDevice dev $ \d -> c_get_key d >>= peekCString

getIp :: TuyaDevice -> IO String
getIp dev = withDevice dev $ \d -> c_get_ip d >>= peekCString

-- ---------------------------------------------------------------------------
-- Connection
-- ---------------------------------------------------------------------------

connect :: TuyaDevice -> String -> IO Bool
connect dev host = withDevice dev $ \d ->
    withCString host $ fmap toBool . c_connect d

disconnect :: TuyaDevice -> IO ()
disconnect dev = withDevice dev c_disconnect

isConnected :: TuyaDevice -> IO Bool
isConnected dev = withDevice dev $ fmap toBool . c_is_connected

reconnect :: TuyaDevice -> IO Bool
reconnect dev = withDevice dev $ fmap toBool . c_reconnect

-- ---------------------------------------------------------------------------
-- Retry
-- ---------------------------------------------------------------------------

setRetryLimit :: TuyaDevice -> Int -> IO ()
setRetryLimit dev n = withDevice dev $ \d -> c_set_retry_limit d (fromIntegral n)

setRetryDelay :: TuyaDevice -> Int -> IO ()
setRetryDelay dev ms = withDevice dev $ \d -> c_set_retry_delay d (fromIntegral ms)

getRetryLimit :: TuyaDevice -> IO Int
getRetryLimit dev = withDevice dev $ fmap fromIntegral . c_get_retry_limit

getRetryDelay :: TuyaDevice -> IO Int
getRetryDelay dev = withDevice dev $ fmap fromIntegral . c_get_retry_delay

-- ---------------------------------------------------------------------------
-- Session negotiation
-- ---------------------------------------------------------------------------

negotiateSession :: TuyaDevice -> String -> IO Bool
negotiateSession dev key = withDevice dev $ \d ->
    withCString key $ fmap toBool . c_neg_sess d

negotiateSessionStart :: TuyaDevice -> String -> IO Bool
negotiateSessionStart dev key = withDevice dev $ \d ->
    withCString key $ fmap toBool . c_neg_start d

negotiateSessionFinalize :: TuyaDevice -> Ptr CChar -> Int -> String -> IO Bool
negotiateSessionFinalize dev buf sz key = withDevice dev $ \d ->
    withCString key $ \ck -> fmap toBool $ c_neg_final d buf (fromIntegral sz) ck

-- ---------------------------------------------------------------------------
-- State queries
-- ---------------------------------------------------------------------------

getProtocol :: TuyaDevice -> IO Protocol
getProtocol dev = withDevice dev $ fmap (protoFromInt . fromIntegral) . c_get_proto

getSessionState :: TuyaDevice -> IO SessionState
getSessionState dev = withDevice dev $ fmap (sessFromInt . fromIntegral) . c_get_sess

getSocketState :: TuyaDevice -> IO SocketState
getSocketState dev = withDevice dev $ fmap (sockFromInt . fromIntegral) . c_get_sock

getLastError :: TuyaDevice -> IO Int
getLastError dev = withDevice dev $ fmap fromIntegral . c_get_err

-- ---------------------------------------------------------------------------
-- Async mode
-- ---------------------------------------------------------------------------

setAsyncMode :: TuyaDevice -> Bool -> IO ()
setAsyncMode dev b = withDevice dev $ \d -> c_set_async d (if b then 1 else 0)

isSocketReadable :: TuyaDevice -> IO Bool
isSocketReadable dev = withDevice dev $ fmap toBool . c_readable

isSocketWritable :: TuyaDevice -> IO Bool
isSocketWritable dev = withDevice dev $ fmap toBool . c_writable

setSessionReady :: TuyaDevice -> IO Bool
setSessionReady dev = withDevice dev $ fmap toBool . c_set_ready

-- ---------------------------------------------------------------------------
-- Low-level message building and decoding
-- ---------------------------------------------------------------------------

buildMessage :: TuyaDevice -> CInt -> String -> String -> IO CInt
buildMessage dev cmd payload key = withDevice dev $ \d ->
    allocaBytes bufSize $ \buf ->
        withCString payload $ \cp ->
        withCString key $ \ck ->
            c_build d buf cmd cp ck

decodeMessage :: TuyaDevice -> Ptr CChar -> Int -> String -> IO (Maybe String)
decodeMessage dev buf sz key = withDevice dev $ \d ->
    withCString key $ \ck -> do
        ptr <- c_decode d buf (fromIntegral sz) ck
        if ptr == nullPtr then return Nothing
        else Just <$> consume ptr

generatePayload :: TuyaDevice -> CInt -> String -> String -> IO (Maybe String)
generatePayload dev cmd devId dps = withDevice dev $ \d ->
    withCString devId $ \cd ->
    withCString dps $ \cdps -> do
        ptr <- c_gen d cmd cd cdps
        if ptr == nullPtr then return Nothing
        else Just <$> consume ptr

send :: TuyaDevice -> Ptr CChar -> Int -> IO Int
send dev buf sz = withDevice dev $ \d ->
    fmap fromIntegral $ c_send d buf (fromIntegral sz)

receive :: TuyaDevice -> Ptr CChar -> Int -> Int -> IO Int
receive dev buf maxsz minsz = withDevice dev $ \d ->
    fmap fromIntegral $ c_receive d buf (fromIntegral maxsz) (fromIntegral minsz)

-- ---------------------------------------------------------------------------
-- High-level round-trip operations
-- ---------------------------------------------------------------------------

setValueBool :: TuyaDevice -> Int -> Bool -> IO String
setValueBool dev dp v = withDevice dev $ \d ->
    consume =<< c_set_bool d (fromIntegral dp) (if v then 1 else 0)

setValueInt :: TuyaDevice -> Int -> Int -> IO String
setValueInt dev dp v = withDevice dev $ \d ->
    consume =<< c_set_int d (fromIntegral dp) (fromIntegral v)

setValueString :: TuyaDevice -> Int -> String -> IO String
setValueString dev dp v = withDevice dev $ \d ->
    withCString v $ \cv -> consume =<< c_set_str d (fromIntegral dp) cv

setValueFloat :: TuyaDevice -> Int -> Double -> IO String
setValueFloat dev dp v = withDevice dev $ \d ->
    consume =<< c_set_float d (fromIntegral dp) (realToFrac v)

turnOn :: TuyaDevice -> Int -> IO String
turnOn dev dp = withDevice dev $ \d -> consume =<< c_turn_on d (fromIntegral dp)

turnOff :: TuyaDevice -> Int -> IO String
turnOff dev dp = withDevice dev $ \d -> consume =<< c_turn_off d (fromIntegral dp)

status :: TuyaDevice -> IO String
status dev = withDevice dev $ \d -> consume =<< c_status d

heartbeat :: TuyaDevice -> IO String
heartbeat dev = withDevice dev $ \d -> consume =<< c_heartbeat d

-- | Type-aware dispatcher: Bool, Int, Double, or String.
data SVal = SBool Bool | SInt Int | SFloat Double | SString String

setValue :: TuyaDevice -> Int -> SVal -> IO String
setValue dev dp (SBool   v) = setValueBool dev dp v
setValue dev dp (SInt    v) = setValueInt  dev dp v
setValue dev dp (SFloat  v) = setValueFloat dev dp v
setValue dev dp (SString v) = setValueString dev dp v

-- ---------------------------------------------------------------------------
-- Device22
-- ---------------------------------------------------------------------------

setDevice22 :: TuyaDevice -> Maybe String -> IO ()
setDevice22 dev Nothing  = withDevice dev $ \d -> c_set_d22 d nullPtr
setDevice22 dev (Just s) = withDevice dev $ \d ->
    withCString s $ c_set_d22 d

isDevice22 :: TuyaDevice -> IO Bool
isDevice22 dev = withDevice dev $ fmap toBool . c_is_d22
